codex/logic_code/versions.tf.md

```diff
@@ -0,0 +1,17 @@
+# `versions.tf` Rules
+
+`version.tf` file can only contain one `terraform` block.
+
+The first line of this `terraform` block should define `required_version = ">= 1.1"`.
+
+`terraform` block should contain a block called `required_providers`, the content of it is the restrictions for provider's version. Provider's version restriction should be sorted in alphabetical order, same for the assignment statements. `version` restrctions should use `>=` if not specified. No other kinds of restrictions can be used.
+
+All the providers used in the module should be defined in `required_providers`.
+
+Since a major version upgrade of the provider can lead to a breaking change, a major version upgrade of the provider must come with a major version upgrade of the module itself.
```

Comment:
Better be:

```markdown
Since a major version upgrade of the provider can lead to a breaking change, a major version upgrade of the provider must come with a major version upgrade of the module itself. Which means, we should almost always declare `azurerm` block like the following example:
```

```hcl
azurerm = {
  source  = "hashicorp/azurerm"
  version = ">= 3.11, < 4.0"
}
```